text/0000-grid-primitives.md

```diff
@@ -0,0 +1,618 @@
+
+```
+    get_schema(schema_name)
+    set_schema(schema_name, schema)
+```
+
+### Schema Addressing in the Merkle-Radix State System
+
+All Grid addresses start with the 6-hex-character namespace prefix "621dee".
+Schemas are further prefixed under the Grid namespace with reserved "01", which
+indicates "Schemas".  The remainder of the address is computed from a portion of
+a SHA-512 hash of the Schema name. 
+
+For example, the address of the "Lightbulb" schema defined in the example above
+is (in Python): 
+
+```
+    "621dee" + "01" + hashlib.sha512("Lightbulb").encode("utf-8")).hexdigest()[:62]
+```
+
+## Transactions
+
+In order to add Schemas to state, a transaction must be used.  The following
+transactions and their execution rules are designed for the Hyperledger Sawtooth
+platform and may differ for other transaction execution platforms.
+
+### Transaction Header
+
+The header for the transactions will include the following:
+
+- `family_name`: `"grid_schema"`
+- `family_version`: `"1.0"`
+- `namespaces`: `[ "621dee" ]`
+
+### Payloads and Execution Rules
+
+#### SchemaPayload
+
+SchemaPayload contains an action enum and the associated action payload.  This
+allows for the action payload to be dispatched to the appropriate logic.
+
+Only the defined actions are available and only one action payload should be
+defined in the SchemaPayload.
+
+```
+message SchemaPayload {
+    enum Actions {
+        UNSET_ACTION = 0;
+        SCHEMA_CREATE = 1;
+        SCHEMA_UPDATE = 2;
+    }
+
+    Action action = 1;
+
+    SchemaCreateAction schema_create = 2;
+    SchemaUpdateAction schema_update = 3
+}
+```
+
+#### SchemaCreateAction
+
+SchemaCreateAction adds a new Schema to state. 
+
+```
+message SchemaCreateAction {
+    string schema_name = 1;
+    string description = 2;
+    repeated PropertyDefinition properties = 10;
+}
+```
+
+The action is validated according to the following rules:
+
+- If a Schema already exists with this name or the name is an empty string, the
+  transaction is invalid.
+- If the property list is empty, the transaction is invalid.
+- The signer of the transaction must be an agent in Pike state and must belong
+  to an organization in Pike state, otherwise the transaction is invalid.
+- The agent must have the permission `can_create_schema` for the organization,
+  otherwise the transaction is invalid.  
+
+The schema is created with the provided fields, in addition to the Pike
+organization ID as the `owner_id`. The schema is then stored in state,
+
+The inputs for SchemaCreateAction must include:
+
+- Address of the Agent submitting the transaction
+- Address of the Schema
+  
+The outputs for SchemaCreateAction must include:
+
+- Address of the Schema  
+
+#### SchemaUpdateAction
+
+SchemaUpdateAction updates a Schema to state. This update only adds new
+Properties to the Schema.
+
+```
+message SchemaUpdateAction {
+    string schema_name = 1;
+    repeated PropertyDefinition properties = 2;
+}
+```
```

Comment:
Versioning seems worth commenting on here.  Existing state doesn't just evolve to new schemas, presumably.

Reply:
As of this RFC, schema's are only additive, which makes all schemas forward-compatible.  The new fields would have to be marked as not required, for this to work.  

Will have to add something about this to rules for updates, and some additional comments about the schemas over time.

Reply:
Added a paragraph to the "Unanswered questions" section.